go-controller/pkg/ovn/gateway_init.go

```diff
@@ -781,3 +787,58 @@ func (oc *DefaultNetworkController) deletePolicyBasedRoutes(policyID, priority s
 
 	return nil
 }
+
+// cleanupStaleMasqueradeData removes following from northbound database
+//   - LogicalRouterStaticRoute for rtoe-<GW_router> OutputPort anf IPPrefix is same as v4 or v6
+//     StaleMasqueradeSubnet
+//   - StaticMACBinding for rtoe-<GW_router> LogicalPort and referencing old DummyNextHopMasqueradeIP
+func deleteStaleMasqueradeResources(nbClient libovsdbclient.Client, routerName string) error {
+	var staleMasqueradeIPs config.MasqueradeIPsConfig
+	var nextHop []net.IP
+	logicalport := types.GWRouterToExtSwitchPrefix + routerName
+	if config.Gateway.V4MasqueradeSubnet != config.Gateway.V4StaleMasqueradeSubnet {
+		if err := config.AllocateV4MasqueradeIPs(iputils.ParseIP(config.Gateway.V4StaleMasqueradeSubnet).ToIP(), &staleMasqueradeIPs); err != nil {
+			return fmt.Errorf("unable to determine stale V4MasqueradeIPs: %s", err)
+		}
+		nextHop = append(nextHop, staleMasqueradeIPs.V4DummyNextHopMasqueradeIP)
+	}
+	if config.Gateway.V6MasqueradeSubnet != config.Gateway.V6StaleMasqueradeSubnet {
+		if err := config.AllocateV6MasqueradeIPs(iputils.ParseIP(config.Gateway.V6StaleMasqueradeSubnet).ToIP(), &staleMasqueradeIPs); err != nil {
+			return fmt.Errorf("unable to determine stale V6MasqueradeIPs: %s", err)
+		}
+		nextHop = append(nextHop, staleMasqueradeIPs.V6DummyNextHopMasqueradeIP)
+	}
+
+	if err := deleteStaleMasqueradeRouteAndMACBinding(nbClient, routerName, nextHop); err != nil {
+		return fmt.Errorf("failed to remove stale MAC binding and static route for logical port %s: %w", logicalport, err)
+	}
+
+	return nil
+}
+
+func deleteStaleMasqueradeRouteAndMACBinding(nbClient libovsdbclient.Client, routerName string, nextHopIPs []net.IP) error {
+	logicalport := types.GWRouterToExtSwitchPrefix + routerName
+	for _, nextHop := range nextHopIPs {
+		prefix := config.Gateway.V4StaleMasqueradeSubnet
+		if utilnet.IsIPv6(nextHop) {
+			prefix = config.Gateway.V6StaleMasqueradeSubnet
+		}
+
+		staticRoutePredicate := func(item *nbdb.LogicalRouterStaticRoute) bool {
+			return item.OutputPort != nil && *item.OutputPort == logicalport &&
+				item.IPPrefix == prefix && utilnet.IPFamilyOfCIDRString(item.IPPrefix) == utilnet.IPFamilyOf(nextHop)
+		}
+		if err := libovsdbops.DeleteLogicalRouterStaticRoutesWithPredicate(nbClient, routerName, staticRoutePredicate); err != nil {
```

Comment:
This function could be refactored to do just one nbdb transaction.
`for _, nextHop := range nextHopIPs {` could be included in the predicates and the Delete operations could be joined.

Reply:
Yes, I agree, delete operations can be joined but it looks cleaner to me now. should I still go ahead and combine both delete operationModels ?

Reply:
Combining all operations into one transaction means there is less chance for a partial cleanup and results in less calls to the db.
 I am ok to keep it as is if that's what you prefer, it shouldn't break anything :)